packages/zoe/src/contracts/autoswap.js

```diff
@@ -55,52 +47,74 @@ export const makeContract = harden((zoe, terms) => {
       const makeInvite = () => {
         const seat = harden({
           swap: () => {
-            let UNITS_IN_INDEX;
-            const amountInFirst = ['offerAtMost', 'wantAtLeast', 'wantAtLeast'];
-            const amountInSecond = [
-              'wantAtLeast',
-              'offerAtMost',
-              'wantAtLeast',
-            ];
-            if (hasValidPayoutRules(amountInFirst, inviteHandle)) {
-              UNITS_IN_INDEX = 0;
-            } else if (hasValidPayoutRules(amountInSecond, inviteHandle)) {
-              UNITS_IN_INDEX = 1;
+            const { proposal } = zoe.getOffer(inviteHandle);
+            const giveTokenA = harden({
+              give: ['TokenA'],
+              want: ['TokenB', 'Liquidity'],
+            });
+            const giveTokenB = harden({
+              give: ['TokenB'],
+              want: ['TokenA', 'Liquidity'],
+            });
+            let giveKeyword;
+            let wantKeyword;
+            if (checkIfProposal(inviteHandle, giveTokenA)) {
+              giveKeyword = 'TokenA';
+              wantKeyword = 'TokenB';
+            } else if (checkIfProposal(inviteHandle, giveTokenB)) {
+              giveKeyword = 'TokenB';
+              wantKeyword = 'TokenA';
             } else {
-              throw rejectOffer(inviteHandle);
+              return rejectOffer(inviteHandle);
             }
-            const UNITS_OUT_INDEX = Nat(1 - UNITS_IN_INDEX);
-            const { newPoolAmountsArray, amountOut } = getPrice(
-              getPoolAmounts(),
-              zoe.getOffer(inviteHandle).amounts[UNITS_IN_INDEX],
-            );
 
```

Comment:
At this point you assume that the Liquidity want in the proposal is empty, likely because it was added by `cleanProposal`. However, you don't actually check that it's empty. If it isn't, the `reallocate` on line 105 will be rejected by Zoe. A well behaved contract should never propose a reallocation to zoe that zoe will reject. There's nothing unsafe here. The rejection will throw before `zoe.complete`. But it is surprising.